File: src/finding_pairs_with_a_certain_sum_1865/solution.py
from collections import Counter
from typing import List


class FindSumPairs:
    def __init__(self, nums1: List[int], nums2: List[int]):
        self._nums1 = sorted(nums1)
        self._nums2 = nums2
        self._nums2ToCount = Counter(nums2)

    def add(self, index: int, val: int) -> None:
        self._nums2ToCount[self._nums2[index]] -= 1
        self._nums2[index] += val
        self._nums2ToCount[self._nums2[index]] += 1

    def count(self, tot: int) -> int:
        result = 0

        for num1 in self._nums1:
            if num1 >= tot:
                break

            result += self._nums2ToCount[tot - num1]

        return result


def test():
    obj = FindSumPairs([1, 1, 2, 2, 2, 3], [1, 4, 5, 2, 5, 4])
    print(obj.count(7))
    obj.add(3, 2)
    print(obj.count(8))
    print(obj.count(4))
    obj.add(0, 1)
    obj.add(1, 1)
    print(obj.count(7))

File: src/finding_pairs_with_a_certain_sum_1865/solution.rs
use std::collections::HashMap;

struct FindSumPairs {
    nums1: Vec<i32>,
    nums2: Vec<i32>,
    nums2_to_count: HashMap<i32, i32>,
}

impl FindSumPairs {
    fn new(mut nums1: Vec<i32>, nums2: Vec<i32>) -> Self {
        nums1.sort();

        let mut nums2_to_count = HashMap::new();
        for &num in &nums2 {
            *nums2_to_count.entry(num).or_insert(0) += 1;
        }

        FindSumPairs {
            nums1,
            nums2,
            nums2_to_count,
        }
    }

    fn add(&mut self, index: i32, val: i32) {
        let index = index as usize;
        self.nums2_to_count
            .entry(self.nums2[index])
            .and_modify(|e| *e -= 1);
        self.nums2[index] += val;
        *self.nums2_to_count.entry(self.nums2[index]).or_insert(0) += 1;
    }

    fn count(&self, tot: i32) -> i32 {
        let mut result = 0;

        for &num1 in &self.nums1 {
            if num1 >= tot {
                break;
            }

            if let Some(&count) = self.nums2_to_count.get(&(tot - num1)) {
                result += count;
            }
        }

        result
    }
}

pub fn test() {
    let mut obj = FindSumPairs::new(vec![1, 1, 2, 2, 2, 3], vec![1, 4, 5, 2, 5, 4]);
    println!("{}", obj.count(7));
    obj.add(3, 2);
    println!("{}", obj.count(8));
    println!("{}", obj.count(4));
    obj.add(0, 1);
    obj.add(1, 1);
    println!("{}", obj.count(7));
}
